src/v/storage/disk_log_impl.cc

```diff
@@ -1731,6 +1735,520 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+    for (; it < _segs.end(); it++) {
+        if (it->get()->is_closed()) {
+            co_return std::nullopt;
+        }
+        if (*it == first_segment) {
+            vlog(
+              stlog.debug,
+              "First offset {} located at {}, offset range size: {}, Segment "
+              "offsets: {}",
+              first,
+              first_segment_file_pos,
+              first_segment_size - first_segment_file_pos,
+              first_segment_offsets);
+            current_size += first_segment_size - first_segment_file_pos;
+        } else {
+            auto sz = it->get()->file_size();
+            current_size += sz;
+            vlog(
+              stlog.debug,
+              "Adding {} bytes to the offset range. Segment offsets: {}, "
+              "current_size: {}",
+              sz,
+              it->get()->offsets(),
+              current_size);
+        }
+        num_segments++;
+        if (current_size > target.target_size) {
+            // Segment size overshoots and we need to use segment index
+            // to find the end offset and its file pos.
+            // We accumulated enough segments to satisfy the query. There're
+            // few cases:
+            // - The beginning and the end of the range is inside the same
+            // segment. The 'num_segments' will be equal to one if this is the
+            // case.
+            // - The beginning and the end are located in different
+            // segments. In both cases we need to find the finish line. The
+            // size calculation is affected by these two possibilities.
+            vlog(
+              stlog.debug,
+              "Offset range size overshoot by {}, current segment size "
+              "{}, current offset range size: {}",
+              current_size - target.target_size,
+              it->get()->file_size(),
+              current_size);
+
+            size_t truncate_after = 0;
+            if (num_segments == 1) {
+                // There are two cases here.
+                // 1. The segment has at least target_size bytes after
+                // base_file_pos,
+                // 2. The segment is too small. In this case we need to clamp
+                // the result.
+                truncate_after = first_segment_file_pos + target.target_size;
+                truncate_after = std::clamp(
+                  truncate_after,
+                  first_segment_file_pos,
+                  it->get()->file_size());
+            } else {
+                // In this case we need to find the truncation point
+                // always starting from the beginning of the segment.
+                //
+                // prev is guaranteed to be smaller than target_size
+                // because we reached this branch.
+                auto prev = current_size - it->get()->file_size();
+                auto delta = target.target_size - prev;
+                truncate_after = delta;
+            }
```

Comment:
I'm a little confused about why this has to be two distinct cases. In either case, isn't the idea to remove `current_size - target.target_size` bytes? And aren't we guaranteed that it's this segment that pushed us over the top?

Would something like this work:

```
auto cur_seg_pos = it->get()->file_size(); // Above calculation `current_size` has assumed that we've accepted to the end of the file, and now we need to truncate.
auto truncate_by = current_size - target.target_size;
auto truncate_after = cur_seg_pos - truncate_by;
```

Reply:
>we need to use the size of the truncated prefix to locate the end of the offset range.

Right, but doesn't `current_size` already account for the missing (or not) prefix?

I suspect these are the same, translating my suggestion, ignoring the clamp:

case of 1 segment:
cur_seg_pos = first_seg_file_size
current_size = first_seg_file_size - first_seg_file_pos
truncate_by = first_seg_file_size - first_seg_file_pos - target.target_size
truncate_after = first_seg_file_size - (first_seg_file_size - first_seg_file_pos - target.target_size)
... = first_seg_file_pos + target.target_size (same as L2147, and sure we can clamp after)

case of more than 1 segment:
cur_seg_pos = iter->file_size()
current_size = size of previous segments + iter->file_size()
truncate_by = size of previous segments + iter->file_size() - target.target_size
truncate_after = iter->file_size() - (size of previous segments + iter->file_size() - target.target_size)
... = target.target_size - size of previous segments (same as L2160)

After writing it all out, I'm more convinced these are the same, though it's somewhat a nit. I do think it makes the code more readable though.